### ISSUE TITLE
Custom querystring parser

### ISSUE THREAD

**Issue by delvedor** [fastify/fastify#1268]
## 🚀 Feature Proposal

Currently, we are using the Node.js's core query string parser, it would be nice to expose an option to support a custom parser, such as [`qs`](https://www.npmjs.com/package/qs).

## Motivation

One of the main design goals of Fastify is *flexibility*, and since we do support a custom body parser, we should support a custom query string parser as well.

## Example

```js
const qs = require('qs')
const fastify = require('fastify')({
  querystring: qs.parse
})
```

**Comment by StarpTech** [comment]
> One of the main design goals of Fastify is flexibility, and since we do support a custom body parser, we should support a custom query string parser as well.

Different body-parser are needed to support different content but the query string parser has a sole purpose. This looks for me like a 100% great use-case for a plugin.

**Comment by ACXgit** [comment]
Thumbs up for the idea, I hit my nose on this issue today https://github.com/fastify/fastify/issues/726 and having a custom query string parser would solve it without dirty hacks.

**Comment by delvedor** [comment]
> Different body-parser are needed to support different content but the query string parser has a sole purpose. This looks for me like a 100% great use-case for a plugin.
 
@StarpTech the problem is that currently there no way to do that, even with a plugin, unless you do some nasty update via a hook in the request object.

**Comment by jsumners** [comment]
Sounds like a good idea. What about `queryStringParser` for the option name? It’s more descriptive of what is being specified.

**Comment by mcollina** [comment]
I'm kind of -1. I think we should switch to `new URL()` for Fastify v3, and that include automatic query parsing. We would end up doing the work twice.

(`url.parse` will get deprecated in Node 12).

**Comment by StarpTech** [comment]
@mcollina How is it related to the feature to plug&play a custom query parser?

**Comment by delvedor** [comment]
> I'm kind of -1.

Why? I think this would be very useful. We should decide if we want it as a configuration option or expose an API to allow encapsulation.

> (url.parse will get deprecated in Node 12).

Should we think about remove it during Fastify v2 RC phase?

**Comment by mcollina** [comment]
I kind of talked it through with @delvedor. I'm ok with it (I was being paranoid about utf8 security).

**Comment by delvedor** [comment]
Closed in #1270.
